src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/MethodCodeNode.cs

```diff
@@ -83,9 +83,37 @@ protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFact
 
             CodeBasedDependencyAlgorithm.AddDependenciesDueToMethodCodePresence(ref dependencies, factory, _method);
 
+            if (_method.IsPInvoke)
+            {
+                MethodSignature methodSig = _method.Signature;
+                if (methodSig.ReturnType.IsDelegate)
+                {
+                    AddPInvokeDelegateParameterDependencies(ref dependencies, factory, methodSig.ReturnType);
+                }
+
+                for (int i=0; i < methodSig.Length; i++)
+                {
+                    if (methodSig[i].IsDelegate)
+                    {
+                        AddPInvokeDelegateParameterDependencies(ref dependencies, factory, methodSig[i]);
+                    }
+                }
+            }
+
             return dependencies;
         }
 
+        private void AddPInvokeDelegateParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        {
+            if (dependencies == null)
+                dependencies = new DependencyList();
+
+            dependencies.Add(factory.ConstructedTypeSymbol(parameter), "Delegate Marshalling Stub");
```

Comment:
Also, is this part testable? Could you add a test that tests the scenario I mentioned (use a delegate type that nobody allocates in a signature and pass a null)? Or is that just an academical issue that doesn't happen because you e.g. can't pass a null delegate to native? (I'm not an interop person, so I don't know, and I'm too lazy to try myself :)).